backends/fs/android/android-saf-fs.cpp

```diff
@@ -0,0 +1,698 @@
+		LOGE("SAFFSTree::createWriteStream failed");
+
+		env->ExceptionDescribe();
+		env->ExceptionClear();
+
+		return nullptr;
+	}
+
+	if (fd == -1) {
+		return nullptr;
+	}
+
+	FILE *f = fdopen(fd, "w");
+	if (!f) {
+		close(fd);
+		return nullptr;
+	}
+
+	return new StdioStream(f);
+}
+
+bool AndroidSAFFilesystemNode::createDirectory() {
+	assert(_safTree != nullptr);
+
+	if (_safNode) {
+		return _flags & DIRECTORY;
+	}
+
+	assert(_safParent);
+
+	JNIEnv *env = JNI::getEnv();
+
+	jstring name = env->NewStringUTF(_newName.c_str());
+
+	jobject child = env->CallObjectMethod(_safTree, _MID_createDirectory, _safParent, name);
+
+	env->DeleteLocalRef(name);
+
+	if (env->ExceptionCheck()) {
+		LOGE("SAFFSTree::createDirectory failed");
+
+		env->ExceptionDescribe();
+		env->ExceptionClear();
+
+		return false;
+	}
+
+	if (!child) {
+		return false;
+	}
+
+	_safNode = env->NewGlobalRef(child);
+	assert(_safNode);
+
+	env->DeleteLocalRef(child);
+
+	cacheData(true);
+
+	return true;
+}
+
+void AndroidSAFFilesystemNode::cacheData(bool force) {
+	if (_cached && !force) {
+		return;
+	}
+
+	JNIEnv *env = JNI::getEnv();
+
+	_flags = env->GetIntField(_safNode, _FID__flags);
+
+	jobject safParent = env->GetObjectField(_safNode, _FID__parent);
+	if (safParent) {
+		if (_safParent) {
+			env->DeleteGlobalRef(_safParent);
+		}
+		_safParent = env->NewGlobalRef(safParent);
+		assert(_safParent);
+		env->DeleteLocalRef(safParent);
+	}
+
+	if (_safParent == nullptr) {
+		jstring nameObj = (jstring)env->GetObjectField(_safTree, _FID__treeName);
+		const char *nameP = env->GetStringUTFChars(nameObj, 0);
+		if (nameP != 0) {
+			_newName = Common::String(nameP);
+			env->ReleaseStringUTFChars(nameObj, nameP);
+		}
+		env->DeleteLocalRef(nameObj);
+	}
+	_path.clear();
+
+	Common::String workingPath;
+
+	jstring pathObj = (jstring)env->GetObjectField(_safNode, _FID__path);
+	const char *path = env->GetStringUTFChars(pathObj, 0);
+	if (path != 0) {
+		workingPath = Common::String(path);
+		env->ReleaseStringUTFChars(pathObj, path);
+	} else {
+		env->DeleteLocalRef(pathObj);
+		return;
+	}
+	env->DeleteLocalRef(pathObj);
+
+	jstring idObj = (jstring)env->CallObjectMethod(_safTree, _MID_getTreeId);
+	if (env->ExceptionCheck()) {
+		LOGE("SAFFSTree::getTreeId failed");
+
+		env->ExceptionDescribe();
+		env->ExceptionClear();
+
+		env->ReleaseStringUTFChars(pathObj, path);
+		env->DeleteLocalRef(pathObj);
+		return;
+	}
+
+	if (!idObj) {
+		return;
+	}
+
+	const char *id = env->GetStringUTFChars(idObj, 0);
+	if (id != 0) {
+		_path = Common::String::format("%s%s%s", SAF_MOUNT_POINT, id, workingPath.c_str());
+		env->ReleaseStringUTFChars(idObj, id);
+	}
+	env->DeleteLocalRef(idObj);
+
+	_cached = true;
+}
+
+const char AddSAFFakeNode::SAF_ADD_FAKE_PATH[] = "/saf";
+
+AddSAFFakeNode::~AddSAFFakeNode() {
+	delete _proxied;
+}
+
+AbstractFSNode *AddSAFFakeNode::getChild(const Common::String &name) const {
+	// We can't call getChild as it's protected
```

Comment:
Is this a TODO? Or is this implementation of `getChild()` not used by engine code?

Reply:
This is a fake node which is only used in the browser at the root (you can't create this fake node using a path).
It's kind of a hack to allow adding SAF trees without having to modify the GUI.

Normally `getChild` isn't used by the browser because it already used `getChildren`.

I don't plan to change the AbstractFSNode API to make it not protected.

If a change would be needed, it means another way to add the SAF node must be found (don't add hacks over hacks).